### PR TITLE
Added container type metrics

### PR DESCRIPTION
**Pull request by kcmannem**
ATC, TSA 

Each container is now classed as "build" or "check", and the number of containers in each class is measured as a metric with each heartbeat.

### CODE REVIEW

**Comment by ddadlani** [comment]
Is `activeContainers` now basically just `buildContainers` + `checkContainers`? Does it make sense to keep `activeContainers` in the heartbeat?

File: atc/metric/metrics.go [hunk 1]
```diff
@@ -111,24 +111,51 @@ func (event SchedulingJobDuration) Emit(logger lager.Logger) {
...
+			Name:  "build containers",
+			Value: event.BuildContainers,
+			State: EventStateOK,
+			Attributes: map[string]string{
+				"worker": event.WorkerName,
+			},
+		},
+	)
+
+	emit(
```
Comment (cirocosta) [inline atc/metric/metrics.go]: Hey,

Given that for `prometheus` we still need to specially treat the metrics in the emitter, it'd be nice to have a case for check containers https://github.com/concourse/concourse/blob/2c374aaaf6ebc619df4439ca6d21c68a59c77fde/atc/metric/emitter/prometheus.go#L333 (or did I miss something?)

Thx!

**Comment by kcmannem** [comment]
@ddadlani not after #3025, since it'll be creating containers periodically which won't be accounted as check or build.

**Comment by vito** [comment]
@kcmannem Do you have time to come back to this? Looks like we just need to add Prometheus support and have the heartbeater ignore containers that don't have properties?

**Comment by vito** [comment]
Woops, now it needs a rebase. :sweat_smile:

**Comment by kcmannem** [comment]
@vito done

**Comment by cirocosta** [comment]
Hey,

I was trying this PR out and I'm not 100% if it's working as intended.

I've set the following pipeline:

```
resources:
  - name: image
    type: registry-image
    source:
      repository: busybox

jobs:
  - name: job
    plan:
      - get: image
      - task: something
        image: image
        config:
          platform: linux
          run:
            path: true
```

Then I triggered the job and waited for a while (such that the registration time would kick in and the metrics would be updated), but it seems like that doesn't really happen 🤔  

```sh
fly -t local workers --json | jq '.[] | .check_containers'
0
fly -t local containers --json | jq '.[] | select(.type == "check") | .id'
"a19439b8-4833-4462-7181-c1f35916f139"
```

I thought that it could be that `fly` would need to change something, but using `fly curl` reveals the same.

Am I missing something? 🤔  

Thanks!

**Comment by ddadlani** [comment]
@cirocosta we tried the same on our local deployment and we don't see 0, we see null implying that the presenter doesn't allow you to see that info.

```
~/w/concourse (cont_metrics) $ fly -t dev workers --json | jq '.[] | .check_containers'
null
```

**Comment by cirocosta** [comment]
Hey,

Sorry for the super slow feedback on this.

There's an error going on when it comes the time to emit the metrics that is preventing the metric from being exposed:

```sh
{"timestamp":"2019-03-28T14:06:40.474047000Z","level":"error","source":"atc","message":"atc.build-containers-count.emit.failed-to-find-platform-in-event","data":{"error":"expected platform to exist in event.Attributes","session":"1158.1"}}
{"timestamp":"2019-03-28T14:06:40.474197200Z","level":"error","source":"atc","message":"atc.check-containers-count.emit.failed-to-find-platform-in-event","data":{"error":"expected platform to exist in event.Attributes","session":"1159.1"}}
```

which makes sense, as we're not emitting the platform:

```
	emit(
		logger.Session("build-containers-count"),
		Event{
			Name:  "worker build containers",
			Value: event.BuildContainers,
			State: EventStateOK,
			Attributes: map[string]string{
				"worker": event.WorkerName,
			},
		},
	)
```

Thanks!

**Comment by kcmannem** [comment]
@cirocosta I think i fixed it this time. I looked at the output from localhost:9001 and saw the metric being emitted.

**Comment by marco-m** [comment]
Hey @cirocosta in my understanding this is ready for your review, when you have time. Thanks :-)

**Comment by kcmannem** [comment]
closed since, we're reconsidering to have the workers themselves report metrics instead of flowing through atc.

**Comment by marco-m** [comment]
Hello @kcmannem 

> we're reconsidering to have the workers themselves report metrics instead of flowing through atc.

For my understanding: do you mean that you are considering having each worker expose Prometheus metrics ?

Sounds good to me, and more Prometheus-idiomatic.

**Comment by kcmannem** [comment]
Yeah thats why my convo with @cirocosta led to regarding this pr. I'll reopen this if his direction changes.

**Pull request closed without merging**